llvm/lib/Target/ARM/ARMInstrInfo.cpp

```diff
@@ -104,11 +104,12 @@ void ARMInstrInfo::expandLoadStackGuard(MachineBasicBlock::iterator MI) const {
   const GlobalValue *GV =
       cast<GlobalValue>((*MI->memoperands_begin())->getValue());
 
-  if (!Subtarget.useMovt() || Subtarget.isGVInGOT(GV)) {
```

Comment:
`isGVInGOT` is defined as

```
bool ARMSubtarget::isGVInGOT(const GlobalValue *GV) const {
  return isTargetELF() && TM.isPositionIndependent() &&
         !TM.shouldAssumeDSOLocal(*GV->getParent(), GV);
}
```

I believe the function name is misleading and `TM.isPositionIndependent() &&` probably should be removed, so I avoid using it.
However, removing `TM.isPositionIndependent() &&` from the function uncovers other problems, so I do not touch `ARMSubtarget::isGVInGOT`.

Reply:
I agree that we can technically use `LDRLIT_ga_abs` to load the GOT entry address but 
ARMExpandPseudoInsts.cpp:2640 does not use GOT for `LDRLIT_ga_abs`.

Reply:
Technically with dso_preemptable `__stack_chk_guard` with static relocation model, we could emit
```
        ldr     r0, .LCPI0_0   // from LDRLIT_ga_abs
        ldr     r0, [r0]
        ldr     r0, [r0]
...
.LCPI0_0:
        .long   __stack_chk_guard@GOT
```

Unfortunately, such a GOT-generating relocation with computation `GOT(S) + A` does not exist for AArch32.
(The integrated assembler and GNU assembler have an issue that `@GOT` is not rejected)

Therefore, we have to use LDRLIT_ga_pcrel and expand to the PIC+GOT code sequence:
```
.LCPI0_0:
.Ltmp0:
        .long   __stack_chk_guard(GOT_PREL)-((.LPC0_0+8)-.Ltmp0)
```

Reply:
Oh, it's also a relocation thing... that makes sense.  Could you add a comment in the code explaining this so it's clear why we're using this specific path on ELF?

Maybe for clarity, also compute "do we need to force a got-pic sequence" as a boolean, then make the inner check something like `if (TM.isPositionIndependent() || ForceELFGOTPIC`)`.

Reply:
Added a comment.

Edited my previous comment: `R_ARM_GOT_ABS` is present, but it does not have assembler support.